### PR TITLE
[4646] Improve drag & drop usability

### PR DESCRIPTION
**Pull request by rart**
craftercms 4646

### CODE REVIEW

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 1]
```diff
@@ -87,8 +91,13 @@ export function MoveModeZoneMenu(props: MoveModeZoneMenuProps) {
...
 
   const onTrash = () => {
-    contentController.deleteItem(modelId, fieldId, index);
-    onCancel();
+    const minCount = runValidation(findContainerRecord(modelId, fieldId, index).id, 'minCount', [elementIndex - 1]);
+    if (minCount) {
+      post(validationMessage(minCount));
+    } else {
+      contentController.deleteItem(modelId, fieldId, index);
+      onCancel();
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx:102]: Must add confirm.

File: ui/guest/src/react/Guest.tsx [hunk 2]
```diff
@@ -153,6 +153,7 @@ function Guest(props: GuestProps) {
             // Click & dblclick require stopping as early as possible to avoid
             // navigation or other click defaults.
             if (type === 'click' || 'dblclick' === type) {
+              click$.next({ type, event, record });
```
Comment (rart) [inline ui/guest/src/react/Guest.tsx]: Move to click epic?
We may not need this. The epic, based on the state might be able to decide if it dispatches `startListening()`.

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 3]
```diff
@@ -87,8 +91,13 @@ export function MoveModeZoneMenu(props: MoveModeZoneMenuProps) {
   };
 
   const onTrash = () => {
-    contentController.deleteItem(modelId, fieldId, index);
-    onCancel();
+    const minCount = runValidation(findContainerRecord(modelId, fieldId, index).id, 'minCount', [elementIndex - 1]);
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx]: Why `elementIndex - 1`?

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 4]
```diff
@@ -137,14 +146,25 @@ export function MoveModeZoneMenu(props: MoveModeZoneMenuProps) {
...
     };
+
     window.addEventListener('click', onClickingOutsideOfSelectedZone);
     return () => {
+      subscription.unsubscribe();
       window.removeEventListener('click', onClickingOutsideOfSelectedZone);
     };
-  }, []);
+    // Linter doesn't realise that refs is a ref and needn't be on the effect
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx]: Pull & remove

**Review by rart — COMMENTED** [review]
@aachavarria please review these.

File: ui/guest/src/react/MoveModeZoneMenu.tsx (hunk 3 above)
Comment (aachavarria) [reply ui/guest/src/react/MoveModeZoneMenu.tsx]: fixed

File: ui/guest/src/react/MoveModeZoneMenu.tsx (hunk 4 above)
Comment (aachavarria) [reply ui/guest/src/react/MoveModeZoneMenu.tsx]: fixed

File: ui/guest/src/react/Guest.tsx (hunk 2 above)
Comment (aachavarria) [reply ui/guest/src/react/Guest.tsx]: fixed

File: ui/guest/src/store/epics/root.ts [hunk 5]
```diff
@@ -373,6 +372,15 @@ const epic = combineEpics<GuestStandardAction, GuestStandardAction, GuestState>(
               of(setEditingStatus({ status: EditingStatus.FIELD_SELECTED }))
             );
           }
+        } else if (
+          state.status === EditingStatus.FIELD_SELECTED &&
+          state.highlightMode === HighlightMode.MOVE_TARGETS
+        ) {
+          const movableRecordId = iceRegistry.getMovableParentRecord(record.iceIds[0]);
+          if (movableRecordId !== Number(Object.keys(state.highlighted)[0])) {
```
Comment (rart) [inline ui/guest/src/store/epics/root.ts]:
```suggestion
          const movableRecordId = iceRegistry.getMovableParentRecord(record.iceIds[0]);
          if (state.highlighted[movableRecordId] === void 0) {
```

File: ui/guest/src/react/MoveModeZoneMenu.tsx [hunk 6]
```diff
@@ -136,13 +148,14 @@ export function MoveModeZoneMenu(props: MoveModeZoneMenuProps) {
...
-      console.log('click');
+      refs.current.onCancel();
     };
+
     window.addEventListener('click', onClickingOutsideOfSelectedZone);
     return () => {
       window.removeEventListener('click', onClickingOutsideOfSelectedZone);
     };
-  }, []);
+  }, [iceId]);
```
Comment (rart) [inline ui/guest/src/react/MoveModeZoneMenu.tsx:158]: This effect doesn't use iceId. Why listed as dependency?
```suggestion
  }, []);
```

File: ui/guest/src/classes/ContentController.ts [hunk 7]
```diff
@@ -453,6 +481,11 @@ export function sortItem(
   const collection = getCollection(models[modelId], fieldId, currentIndex);
   const result = getCollectionWithoutItemAtIndex(collection, currentIndexParsed);
 
+  // TODO: modelHierarchyMap is not being uploaded after models update
+  // modelHierarchyMap is used on getMovableParentRecord, the one who helps to highlight the zone
+  modelHierarchyMap[collection[currentIndexParsed]].parentContainerFieldIndex = targetIndex;
+  modelHierarchyMap[collection[targetIndex]].parentContainerFieldIndex = currentIndex;
```
Comment (rart) [inline ui/guest/src/classes/ContentController.ts:487]: We just need to (1) add logic to handle dotted notations here (2) consider repeat groups (3) all the items that get displaced based on the position the sort target moved.

**Pull request merged**